src/raftstore/store/debug.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{error, result};
+use kvproto::debugpb::*;
+use rocksdb::DB;
+
+use raftstore::store::keys;
+use raftstore::store::Engines;
+use raftstore::store::engine::Peekable;
+use storage::{CF_DEFAULT, CF_LOCK, CF_RAFT, CF_WRITE};
+
+pub type Result<T> = result::Result<T, Error>;
+
+quick_error!{
+    #[derive(Debug)]
+    pub enum Error {
+        InvalidArgument(msg: String) {
+            description(msg)
+            display("Invalid Argument {:?}", msg)
+        }
+        NotFound(msg: String) {
+            description(msg)
+            display("Not Found {:?}", msg)
+        }
+        Other(err: Box<error::Error + Sync + Send>) {
+            from()
+            cause(err.as_ref())
+            description(err.description())
+            display("{:?}", err)
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Debug {
+    engines: Engines,
+}
+
+impl Debug {
+    pub fn new(engines: Engines) -> Debug {
+        Debug { engines }
+    }
+
+    pub fn get(&self, cf: CF, key_encoded: &[u8]) -> Result<Vec<u8>> {
+        let db = try!(self.get_db(cf));
+        let cf = try!(cf_to_str(cf));
+        match db.get_value_cf(cf, &keys::data_key(key_encoded)) {
+            Ok(Some(v)) => Ok((&v).to_vec()),
+            Ok(None) => Err(Error::NotFound(
+                format!("get none value for encoded key {:?}", key_encoded,),
+            )),
+            Err(e) => Err(box_err!(e)),
+        }
+    }
+
+    fn get_db(&self, cf: CF) -> Result<&DB> {
```

Comment:
This method seems useless.

Reply:
Why useless, it is already used in `get()`, and it will be used in other APIs.

Reply:
Because it returns kv engine for all cfs. Actually we always know which engine to be used in each API.

Reply:
Yes, but it is only for now, it may access both db in the future.

Reply:
In what situation that `get_db` is a requirement?

Reply:
No, but `match` is, `get_db()` is a thin wrapper of `match` so that we only need to write one line of code `try!(get_db())`.

Reply:
In what situation `match` is a requirement? Why can't it just use `self.engines.kv_engine` and `self.engines.raft_engine`?

Reply:
I mean different `CF` requires different DB, for now all `CF`s require `kv_engine`, this may not be the case in the future. Let's say a new `CF:RAFT_DEFAULT` for getting `raft_engine`, and the `GetRequest` may carry this cf, then we have to match `CF`s for the right DB it wants.

Reply:
This can't be true. The structure of keys in `kv_engine` and `raft_engine` are completely different. When access `raft_engine`, it either access meta or raft entry both of which should not be processed via `GetRequest` but via an explicit API.

Besides it's very strange that some keys are prefixed with a 'z' but some are not. It will be more clear and consistent to provide a `get` method to serves data KV only, and provide a `raw_get` to serves any KV without any modification to key. In latter case, an `engine` arg should be passed explicitly.